File: system_design/rate_limiter/python/rate_limiter_token.py
import time
from typing import final


@final
class RateLimiter:
    def __init__(self, capacity: int, refill_rate: int) -> None:
        self.capacity = capacity
        self.refill_rate = refill_rate
        self.tokens = capacity
        self.last_refill_time = time.time()

    def allow_request(self, tokens_requested: int = 1) -> bool:
        now = time.time()

        # Calculate tokens gained since last check
        elapsed = now - self.last_refill_time
        self.tokens = min(self.capacity, self.tokens + (elapsed * self.refill_rate))
        self.last_refill_time = now

        if self.tokens > tokens_requested:
            self.tokens -= tokens_requested
            return True
        return False

File: system_design/rate_limiter/python/go/rate_limiter_token.go
package main

import (
	"time"
)

type RateLimiter struct {
	Capacity       int64
	RefillRate     int64
	Tokens         int64
	LastRefillTime int64
}

func NewRateLimiter(capacity int64, refill_rate int64) *RateLimiter {
	return &RateLimiter{
		Capacity:       capacity,
		RefillRate:     refill_rate,
		Tokens:         capacity,
		LastRefillTime: time.Now().UnixNano(),
	}
}

func (r RateLimiter) AllowRequest(tokens_requested int64) bool {
	now := time.Now().UnixMilli()

	elapsed := now - r.LastRefillTime
	if r.Capacity < r.Tokens+(elapsed*int64(r.RefillRate)) {
		r.Tokens = r.Capacity
	} else {
		r.Tokens = r.Tokens + (elapsed * int64(r.RefillRate))
	}
	r.LastRefillTime = now

	if r.Tokens > tokens_requested {
		r.Tokens -= tokens_requested
		return true
	}
	return false
}
